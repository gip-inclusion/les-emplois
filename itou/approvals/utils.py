from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
