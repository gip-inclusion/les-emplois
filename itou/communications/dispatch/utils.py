class JobSeekerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_job_seeker


class EmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_employer


class PrescriberNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_prescriber


class PrescriberOrEmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and (self.user.is_prescriber or self.user.is_employer)
