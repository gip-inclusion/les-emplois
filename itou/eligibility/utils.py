from collections import Counter


def geiq_allowance_amount(is_authorized_prescriber, administrative_criteria) -> int:
    """Amount of granted allowance for job seeker.

    Calculated in function of:
        - author kind
        - number, annex and level of administrative criteria.
    Currently, only 3 amounts possible:
        - 0
        - 814EUR
        - 1400EUR
    """

    # Even when no criteria in the diagnosis
    if is_authorized_prescriber:
        return 1400

    # Count by annex
    annex_cnt = Counter(c.annex for c in administrative_criteria)

    # Only annex 2 administrative criteria have a level defined
    level_cnt = Counter(c.level for c in administrative_criteria if c.level)

    # At least one level 1 criterion or at least two level 2 criteria
    if level_cnt["1"] > 0 or level_cnt["2"] > 1:
        return 1400

    # Criteria in both annex ("1+2") must be counted as annex 1
    if annex_cnt["1"] + annex_cnt["1+2"] > 0:
        return 814

    return 0
