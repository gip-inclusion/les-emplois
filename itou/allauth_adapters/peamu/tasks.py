import httpx
from django.conf import settings
from django.utils.http import urlencode
from huey.contrib.djhuey import task


@task()
def huey_logout_from_pe_connect(peamu_id_token, hp_url):
    params = {"id_token_hint": peamu_id_token, "redirect_uri": hp_url}
    peamu_logout_url = f"{settings.PEAMU_AUTH_BASE_URL}/compte/deconnexion?{urlencode(params)}"
    return httpx.get(peamu_logout_url)
