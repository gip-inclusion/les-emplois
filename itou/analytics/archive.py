from django.db.models import Count, Q
from django.utils import timezone

from itou.analytics.models import DatumCode
from itou.archive.management.commands.notify_inactive_jobseekers import (
    inactive_jobseekers_without_recent_related_objects,
)
from itou.archive.models import (
    AnonymizedApplication,
    AnonymizedApproval,
    AnonymizedCancelledApproval,
    AnonymizedGEIQEligibilityDiagnosis,
    AnonymizedJobSeeker,
    AnonymizedProfessional,
    AnonymizedSIAEEligibilityDiagnosis,
)
from itou.users.models import User, UserKind
from itou.utils.constants import INACTIVITY_PERIOD


def collect_archive_data():
    inactive_since = timezone.now() - INACTIVITY_PERIOD
    counts = User.objects.aggregate(
        anonymized_professionals_not_deleted=Count(
            "pk",
            filter=Q(
                kind__in=UserKind.professionals(), upcoming_deletion_notified_at__isnull=False, email__isnull=True
            ),
        ),
        notified_job_seekers=Count(
            "pk",
            filter=Q(kind=UserKind.JOB_SEEKER, upcoming_deletion_notified_at__isnull=False),
        ),
        notified_professionals=Count(
            "pk",
            filter=Q(
                kind__in=UserKind.professionals(),
                upcoming_deletion_notified_at__isnull=False,
                email__isnull=False,
            ),
        ),
        notifiable_professionals=Count(
            "pk",
            filter=Q(
                kind__in=UserKind.professionals(),
                upcoming_deletion_notified_at__isnull=True,
                last_login__lt=inactive_since,
            ),
        ),
    )

    return {
        DatumCode.ANONYMIZED_PROFESSIONALS_DELETED: AnonymizedProfessional.objects.count(),
        DatumCode.ANONYMIZED_PROFESSIONALS_NOT_DELETED: counts["anonymized_professionals_not_deleted"],
        DatumCode.ANONYMIZED_JOB_SEEKERS: AnonymizedJobSeeker.objects.count(),
        DatumCode.ANONYMIZED_APPLICATIONS: AnonymizedApplication.objects.count(),
        DatumCode.ANONYMIZED_APPROVALS: AnonymizedApproval.objects.count(),
        DatumCode.ANONYMIZED_CANCELLED_APPROVALS: AnonymizedCancelledApproval.objects.count(),
        DatumCode.ANONYMIZED_IAE_ELIGIBILITY_DIAGNOSIS: AnonymizedSIAEEligibilityDiagnosis.objects.count(),
        DatumCode.ANONYMIZED_GEIQ_ELIGIBILITY_DIAGNOSIS: AnonymizedGEIQEligibilityDiagnosis.objects.count(),
        DatumCode.NOTIFIED_PROFESSIONALS: counts["notified_professionals"],
        DatumCode.NOTIFIED_JOB_SEEKERS: counts["notified_job_seekers"],
        DatumCode.NOTIFIABLE_PROFESSIONALS: counts["notifiable_professionals"],
        DatumCode.NOTIFIABLE_JOB_SEEKERS: inactive_jobseekers_without_recent_related_objects(
            inactive_since=inactive_since, notified=False
        ).count(),
    }
