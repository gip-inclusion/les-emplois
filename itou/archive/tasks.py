import logging

from huey.contrib.djhuey import task

from itou.utils.brevo import BrevoClient


logger = logging.getLogger(__name__)


@task(retries=24 * 6 * 90, retry_delay=10 * 60, context=True)  # Retry every 10 minutes during 90 days.
def async_delete_contact(email, *, task=None):
    with BrevoClient() as brevo_client:
        if task.retries % 100 == 0:
            logger.warning("Attempting to delete email %s, remaining %d retries", email, task.retries)
        brevo_client.delete_contact(email)
