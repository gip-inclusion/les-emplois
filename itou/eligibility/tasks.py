import logging
from json import JSONDecodeError

import httpx
from django.apps import apps
from django.utils import timezone
from huey.contrib.djhuey import task
from huey.exceptions import RetryTask

from itou.eligibility.enums import CERTIFIABLE_ADMINISTRATIVE_CRITERIA_KINDS, AdministrativeCriteriaKind
from itou.utils.apis import api_particulier
from itou.utils.types import InclusiveDateRange


logger = logging.getLogger("APIParticulierClient")


def certify_criteria(eligibility_diagnosis):
    job_seeker = eligibility_diagnosis.job_seeker
    if not api_particulier.has_required_info(job_seeker):
        logger.info("Skipping {job_seeker.pk=}, missing required information.")
        return
    SelectedAdministrativeCriteria = eligibility_diagnosis.administrative_criteria.through
    criteria = (
        SelectedAdministrativeCriteria.objects.filter(
            administrative_criteria__kind__in=CERTIFIABLE_ADMINISTRATIVE_CRITERIA_KINDS,
            eligibility_diagnosis=eligibility_diagnosis,
        )
        .select_for_update(no_key=True)
        .select_related("administrative_criteria")
    )
    with api_particulier.client() as client:
        for criterion in criteria:
            # Only the RSA criterion is certifiable at the moment,
            # but this may change soon with the addition of `parent isolé` and `allocation adulte handicapé`.
            if criterion.administrative_criteria.kind == AdministrativeCriteriaKind.RSA:
                try:
                    data = api_particulier.revenu_solidarite_active(client, job_seeker)
                except httpx.HTTPStatusError as exc:
                    criterion.data_returned_by_api = exc.response.json()
                    logger.error(
                        "Error certifying criterion %r: code=%d json=%s",
                        criterion,
                        exc.response.status_code,
                        criterion.data_returned_by_api,
                    )
                    match exc.response.status_code:
                        case 400 | 404:
                            # Job seeker not found or missing profile information.
                            criterion.certified_at = timezone.now()
                        case 429:
                            # https://particulier.api.gouv.fr/developpeurs#respecter-la-volumétrie
                            raise RetryTask(delay=int(exc.response.headers["Retry-After"])) from exc
                        case 503:
                            # TODO: Use the error code instead the message when switching to API v3.
                            if criterion.data_returned_by_api["message"] == (
                                "Erreur de fournisseur de donnée : "
                                "Trop de requêtes effectuées, veuillez réessayer plus tard."
                            ):
                                # The data provider for API particulier returned a 429.
                                # Let’s hope the data provider rate limit has been reset by then.
                                raise RetryTask(delay=3600) from exc
                            else:
                                # The data provider for API particulier likely returned a 500.
                                # According to the API particulier team, it often means integrity
                                # errors on the beneficiary case. The data provider itself aggregates data
                                # from other providers (e.g. regional information systems). When the data
                                # sources aren’t consistent with each other, the data provider cannot
                                # answer.
                                # Retrying won’t fix the issue, and the error has been logged already.
                                pass
                        case _:
                            raise
                else:
                    criterion.certified = data["is_certified"]
                    criterion.certified_at = timezone.now()
                    criterion.data_returned_by_api = data["raw_response"]
                    criterion.certification_period = None
                    start_at, end_at = data["start_at"], data["end_at"]
                    if start_at and end_at:
                        criterion.certification_period = InclusiveDateRange(start_at, end_at)
    SelectedAdministrativeCriteria.objects.bulk_update(
        criteria,
        fields=[
            "certification_period",
            "certified",
            "certified_at",
            "data_returned_by_api",
        ],
    )


@task(retries=24 * 6, retry_delay=10 * 60)  # Retry every 10 minutes for 24h.
def async_certify_criteria(model_name, eligibility_diagnosis_pk):
    model = apps.get_model("eligibility", model_name)
    eligibility_diagnosis = model.objects.select_related("job_seeker__jobseeker_profile").get(
        pk=eligibility_diagnosis_pk
    )
    try:
        certify_criteria(eligibility_diagnosis)
    except (
        httpx.HTTPError,  # Could not connect, unexpected status code, …
        JSONDecodeError,  # Response was not JSON (text, HTML, …).
        RetryTask,  # Rate limiting.
    ):
        # Worth retrying.
        raise
    except Exception as e:
        logger.exception(e)
