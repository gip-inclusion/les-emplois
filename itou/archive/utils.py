import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
