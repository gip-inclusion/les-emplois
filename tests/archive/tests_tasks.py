import httpx
import pytest
from django.conf import settings

from itou.archive.tasks import async_delete_contact
from tests.utils.test_brevo import brevo_client_fixture  # noqa: F401


@pytest.mark.parametrize("status_code", [204, 404])
def test_delete_contact(respx_mock, django_capture_on_commit_callbacks, status_code, caplog, brevo_client):
    email = "somebody@mail.com"
    respx_mock.delete(f"{settings.BREVO_API_URL}/contacts/{email}?identifierType=email_id").mock(
        return_value=httpx.Response(status_code=status_code)
    )

    with django_capture_on_commit_callbacks(execute=True):
        async_delete_contact(email)

    assert [record.levelname for record in caplog.records] == ["INFO"]


@pytest.mark.parametrize("retries", [1, 100, 200])
def test_async_delete_contact_retries_warning(
    respx_mock, django_capture_on_commit_callbacks, retries, caplog, snapshot, brevo_client
):
    email = "somebody@email.com"
    respx_mock.delete(f"{settings.BREVO_API_URL}/contacts/{email}?identifierType=email_id").mock(
        return_value=httpx.Response(status_code=503)
    )

    with django_capture_on_commit_callbacks(execute=True):
        async_delete_contact(email, retries=retries)

    assert respx_mock.calls.called

    if retries % 100 == 0:
        warning_record = next(record for record in caplog.records if record.levelname == "WARNING")
        assert warning_record.message == snapshot(name=f"attempting-to-delete-email-{retries}-retries")


def test_async_delete_contact_does_not_send_HTTP_request(respx_mock, django_capture_on_commit_callbacks, brevo_client):
    for email, called in [(None, False), ("somebody@email.com_old", False), ("somebody@email.com", True)]:
        respx_mock.delete(f"{settings.BREVO_API_URL}/contacts/{email}?identifierType=email_id").mock(
            return_value=httpx.Response(status_code=204)
        )
        with django_capture_on_commit_callbacks(execute=True):
            async_delete_contact(email)
        assert respx_mock.calls.called is called
