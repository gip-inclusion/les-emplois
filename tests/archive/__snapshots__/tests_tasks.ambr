# serializer version: 1
# name: test_async_delete_contact_retries_warning[100][attempting-to-delete-email-100-retries]
  'Attempting to delete email somebody@email.com, remaining 100 retries'
# ---
# name: test_async_delete_contact_retries_warning[200][attempting-to-delete-email-200-retries]
  'Attempting to delete email somebody@email.com, remaining 200 retries'
# ---
